README.md
SVN Access Manager Readme
=========================

SVN Access Manager is a powerful tool for managing access to subversion 
repositories. The tool provides user and group management and access 
rights (read/write) to dedicated paths in a repository as well.

Subversion grants access rights recursively. That means that creating
access rights to repositories is not as complex as it seams.

Let's have a small example.

Repository testrepo:

trunk  
      /dir1  
      /dir2  
          /dir3  
          /dir4  
      /dir5  

Let's assume that there're two users defined for this repository, user A and
user B. User A should have read access to the whole repository but should only
be allowed to write into "dir2" without the sub directories. User B should have
write access to the whole repository except "dir5". For dir5 he should not have
any access.

You can set the access rights with SVN Access Manager according to this created
access file:

[testrepo:/trunk/]  
A = r  
B = rw  
 
[testrepo:/trunk/dir2]  
A = rw  
 
[testrepo:/trunk/dir2/dir3]  
A =  
 
[testrepo:/trunk/dir2/dir4]  
A =  
 
[testrepo:/trunk/dir5]  
B =  
 


Requirements
============

To use SVN Access Manager you need

- a working subversion repository
- Apache 2.0 or higher
- PHP 4 or higher with php mysql and php ldap,
  recommended is PHP 5.3
- Mysql 5
- a web browser supporting HTML 5



seLinux
========

In the doc directory there's a folder selinux. There you find a selinux context
for the SVN Access Manager. You can use this context. If you want to create
repositories from within the SVN Access Manager you must add the rules to
allow the webserver to write to your repository directory.

To add the rules you can do as root:
> semodule -i svnaccessmanager.pp



Installation
============

For installation please look into the documentation in the html directory or
simply install the RPM provided on 
https://www.svn-access-manager.org/



Mailinglists
============

There's a mailinglist for announcements of new releases of SVN Access Manager.
The list requires a subscription which can be done here.

New releases are announced on freshmeat.net and sourceforge.net as well.



Credits
=======

SVN Access Manager uses JQuery, JQuery UI, Chosen, a Select Box Enhancer and 
Aria Tables. 

Thanx to Maik, Tobias and Jan for a lot of testing and providing patches
to fix some bugs.



License
=======

SVN Access Manager is distributed under the GPL v2. You find a copy
of this license in this folder.



Warranty
========

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.





Changelog.md
Change Log
==========

Release 0.6.0.12
    - fixed db query, add quotes arround db field with name "recursive"

Release 0.6.0.11
    - fixed bug in user list showing all users as admin users
    - fixed typo in html template
    - fixed user administration, password change uses now password policy of user instead
      of asdmin policy

Release 0.6.0.9
    - fixed bug #57, installer uses a wrong database name for creating messages table.
      Thanks to Włodzimierz Donatowicz
      
Release 0.6.0.8
    - added phpunit tests and a small database xml dump for database phpunit tests
    - added mysqli driver to installer
    - basic tests with php 7
    - added docker container build for testing purposes
    - fixed bug #56, missing default values and projects now need a description
    - fixed email checking, thanks to Yusuf Tambag for the bug report and the patch
    - splitted database functions in installer, moved them into a file for MySQL, PostgreSQL and Oracle
    - decreased complexity of installer code
    - removed duplicated code
    - use Bootstrap 3
    - removed jQuery-UI
    - updated jQuery version
    - redesigned installer and other forms
    - html documentation was removed and can be found here: https://www.svn-access-manager.org/
    - compressed css stylesheets 
    - removed unused files 
    - use constants instead of literals
    - table page sizes will be stored in browser's local storage for each table if possible

Release 0.5.6.6
    - updated selinux policy for CentOS 6 and CentOS 7
    - fixed a typo in the docs
    - fixed a typo in the install.php file
    
Release 0.5.6.5
	- fix for issue #54 Creation of group in authz file doesnt check deleted user.
	  Thanks to Michael Melin
	
Release 0.5.6.4
	- fix for filenames and pathes with blanks. Thanks to David Helwig for the patch.
	
Release 0.5.6.3
	- fixed 2 bugs
	  Thanks to Juergen Prusseit for the patches.
	  
Release 0.5.6.2
	- fixed a typo in the installer
	
	- non mysql database help texts
	
	- support for Apache MD5 password hashing
	  thanks to Bill Meine for the patch
	
Release 0.5.5.22
	- fixed bug ID: 3602643
	  'temporarily loss of "edit access rights" permission'

	- changed select box type at user add from ldap

Release 0.5.5.20
	- fixed bug ID: 3594777
	  'Undefined variable' if search string wasn't found
	  
	- fixed bug ID: 3594781
	  'Undefined index' on project searchresult page
	  
Release 0.5.5.19
	- fixed bug ID: 3585176
	  editable custom fields on general page
	  
	- fixed bug ID: 3589118
	  cron scripts: duplicate require "config.inc.php"
	  
	- fixed bug ID: 3593675
	  Buttons on searchresult page not working
	  
	- fixed bug ID: 3593676
	  wrong date sorting on log page
	  
	- removed unnecessary error_log calls
	  
Release 0.5.5.18
	- changed search in access right list to a filter
	
	- Fixed bug ID: 3585176
	  editable custom fields on general page
	  
	- Feature request ID: 3584758 	
	  select group adm & project resp. via AJAX
	  
	- Feature request ID: 3585100
	  Ajax based search for list_access_rights.php
	  
	- Feature request ID: 3585094
	  ajax based search in reports
	  
	- Feature request ID: 3586321
	  replace show user/group select box
	  
	- table filters do not start compare from the beginning 
	  of the table cell content
	  
	- new updated help texts. Thanks to Jan for the new verion
	  of the help texts.
	  
Release 0.5.5.17
	- added userid to groupadmin search
	  
	- show only groups an user may work with in search results
	 
	- write permission violations to log
	
	- added search to repository access rights (Repository, path)
	
	- Fixed bug ID: 3584773
	  missing $t-Variables on general page
	  
	- Fixed bug ID: 3581934
	  Database error when deleting project
	 
Release 0.5.5.16
	- Feature request ID: 3582680
	  allow every user a "show my permissions" report
	  added a permission overview to the general view
	  
	- Fixed an issue with administration rights
	
Release 0.5.5.15
	- improved documentation
	
	- documentation is now available as link
	
	
Release 0.5.5.14
	- added some help texts
	
	- Fixed Bug ID: 3580163
	  Missing results in user report
	  
	  
Release 0.5.5.13
	- redesigned help system, added help texts
	
	- jump back to repository list if repo is created or 
	  modified successfully
	
	- create an error message if ldap is not reachable and 
	  do not loop in login
	  
	- small selinux documentation in the README file
	
	- Feature request ID: 3579139
	  additional search fields
	  
	- bulk add functionality for LDAP users
	
	
Release 0.5.5.12
	- Bug fix in rep_show_group.php
	
	- removed old, unneeded files and folders
	
	- changed session handling to be compatible with PHP 5.4
	  
	- renamed hex2bin function to myhex2bin to avoid conflicts 
	  with PHP 5.4
	  
	- Bug Fix in db-functions-adodb.inc.php
	
	- some typos corrected in German language file
	
	- added tool tips
	
	- Fixed Bug ID: 3577838
	  installer doesn't set new ldap options
	  
	  
Release 0.5.5.11
	- Feature request ID: 3576275
	  Encoding LDAP name and given name to handle German umlauts
	  
	- Feature request ID: 3576273
	  Sorting LDAP Users
	  
	- Feature request ID: 3576346
	  Writing root access rights
	  
	- Feature request ID: 3576692
	  Report "show group"
	  
	- Bug Fix in LDAP connect, wrong protocol version was used
	
	- Bug Fix in user and group report
	  output error message if no user or no group are selected
	  
	- added access by column to user report
	
	- Feature request ID: 3576279
	  Bind to LDAP/AD with given user credentials. This feature
	  is only needed for Active Directory at the moment.
	
	
	Thanks to Tobias for sending a lot of patches
	
	  
Release 0.5.5.10
	- Feature request ID: 3575716
	  checkSession.php with Anti-Cache Parameter
	  
	- Bug fix ID: 3575725
	  field order list_users.tpl wrong when using ldap
	  
	- Bug fix ID: 3575798
	  $schema is not set is fixed
	  searching by name and given name or vice versa is now possible
	  

docker/files/config.inc.php
<?php

/*
    SVN Access Manager - a subversion access rights management tool
    Copyright (C) 2008 Thomas Krieger <tom@svn-access-manager.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*

File:            config.inc.php.tpl
$LastChangedDate: 2014-01-01 21:10:25 +0100 (Wed, 01 Jan 2014) $
$LastChangedBy: tom_krieger $

$Id: config.inc.php.tpl 352 2014-01-01 20:10:25Z tom_krieger $

*/


if (preg_match ("/config.inc.php/", $_SERVER['PHP_SELF'])) {

   header ("Location: login.php");
   exit;

}


$CONF['install_base']                   = '/usr/share/svn-access-manager';

// Language config
$CONF['default_language']               = 'en';
$CONF['default_locale']                 = 'en_US';
$CONF['supported_languages']    = array ('de', 'de_DE', 'en', 'en_US');

// Database Config
$CONF['database_type']                  = 'mysqli';
$CONF['database_host']                  = '###DBHOST###';
$CONF['database_user']                  = '###DBUSER###';
$CONF['database_password']              = '###DBPASS###';
$CONF['database_name']                  = '###DBNAME###';
$CONF['database_prefix']                = '';
$CONF['database_innodb']                = 'YES';
$CONF['database_charset']               = '###CHARSET###';
$CONF['database_collation']             = '###COLLATION###';
$CONF['database_schema']                = '';
$CONF['database_tablespace']    = '';
$CONF['session_in_db']                  = 'YES';

$CONF['website_charset']                = 'iso-8859-15';
$CONF['website_url']                    = 'http://localhost:7280/';

// Sort order for user
$CONF['user_sort_fields']               = "name,givenname";
$CONF['user_sort_order']                = "ASC";

// Site Admin
// Define the Site Admins email address below.
$CONF['admin_email']                    = 'svn-admin@example.com';

$CONF['encrypt']                                = 'system';
$CONF['generate_password']              = 'YES';

// Lost password
$CONF['lostPwSender']                   = 'admin@example.com';
$CONF['lostPwMaxError']                 = 3;
$CONF['lostPwLinkValid']                = 2;

$CONF['logging']                                = 'YES';

// Page Size
// Set the number of entries that you would like to see
// in one page.
$CONF[PAGESIZE]                              = '50';

$CONF['passwordSpecialChars']   = '[\!\"\§\$\%\/\(\)=\?\*\+\#\-\_\.\:\,\;\<\>\|\@]';
$CONF['passwordSpecialCharsTxt']= '!"§$%/()=?*+#-_.:,;<>|@';
$CONF['minPasswordlength']              = 14;
$CONF['minPasswordlengthUser']  = 8;
// Set passwword complexity. A password must consist of four different goups och charactsers.
// The four groups are locer-case and upper-case characters, special characters and digits
// The two config varaibles $CONF['minPasswordGroups'] and $CONF['minPasswordGroupsUser']
// allow to set how many groups a password must consist of. The maximum is 4, the minimum is 1!
// If the values are out of range, for user 3 is assumed and for admins 4 is assumed
$CONF['minPasswordGroups']              = 4;
$CONF['minPasswordGroupsUser']  = 3;

// password expiry stuff
$CONF['password_expires']               = 60;
$CONF['password_expires_warn']  = 50;
// define default value for password expire, allowed values 0 and 1
$CONF['expire_password']                = '1';

# valid values for pwcrypt are sha, apr-md5, md5, or crypt
$CONF['pwcrypt']                                = 'apr-md5';

$CONF['copyright']                              = '(C) 2008-2018 Thomas Krieger (tom(at)svn-access-manager(dot)org)';

$CONF['svn_command']                    = '/usr/bin/svn';
$CONF['svnadmin_command']               = '/usr/bin/svnadmin';
$CONF['repo_compatibility']             = '--pre-1.6-compatible';
$CONF['grep_command']                   = '/usr/bin/grep';
$CONF['use_javascript']                 = 'YES';

$CONF['SVNAccessFile']                  = '/etc/svn/svnaccess';
$CONF['AuthUserFile']                   = '/etc/svn/svnpasswd';
$CONF['ViewvcConf']                             = '/etc/svn/viewvc-apache.conf';
$CONF['ViewvcGroups']                   = '/etc/svn/viewvc-groups';
# ViewVC Alias in the Apache Webserver without trailing /
$CONF['ViewvcLocation']                 = '/viewvc';
$CONF['ViewvcApacheReload']             = 'sudo /usr/bin/systemctl reload httpd';
$CONF['ViewvcRealm']                    = 'ViewVC';
$CONF['createAccessFile']               = 'YES';
$CONF['createUserFile']                 = 'YES';
$CONF['createViewvcConf']               = 'YES';
$CONF['separateFilesPerRepo']   = 'NO';
//Limit access control to directories only, change to files if you want to have
//files listed during access rights management too
//Valid values are "dirs" or "files"
$CONF['accessControl']                  = 'dirs';
// Set users default access right
// valid values are; read, write
$CONF['userDefaultAccess']              = 'read';
// annonymous access option, allowed values are 0 or 1
$CONF['write_anonymous_access_rights']   = 0;

// SVN access file repository path sort order, allowed values are ASC or DESC
$CONF['repoPathSortOrder']              = 'ASC';

// Custom Fields
$CONF['column_custom1']                 = NULL;
$CONF['column_custom2']                 = NULL;
$CONF['column_custom3']                 = NULL;

$CONF['mail_password_warn']             = <<<EOM

Dear %s,

your password for SVN Access Manager is about to expire. Please goto %s, log in and change your password.

Please keep in mind that your account will be locked out automatically if your password was not changed.

Users are locked out if the password was not changed for %s days!

Kind regrads

SVN Access Manager
Administrator

EOM;

$CONF['mail_user_locked']               = <<<EOM

Dear %s,

you account at SVN Access Manager was locked. You did not change your password for %s days.

You can not access the subversion repositories any more. To get access please log into your account at %s and change your password.

Please give about %s minutes until your account is unlocked and you can access the subversion repositories again.

Kind regards

SVN Access Manager
Administrator

EOM;

//LDAP stuff
$CONF['use_ldap']                               = 'NO';
$CONF['bind_dn']                                = '';
$CONF['bind_pw']                                = '';
$CONF['user_dn']                                = '';
$CONF['user_filter_attr']                       = '';
$CONF['user_objectclass']                       = '';
$CONF['additional_user_filter']                 = '';
$CONF['ldap_server']                            = '';
$CONF['ldap_port']                              = '389';
$CONF['ldap_protocol']                          = '3';
$CONF['attr_mapping']['uid']                    = 'uid';
$CONF['attr_mapping']['name']                   = 'sn';
$CONF['attr_mapping']['givenName']              = 'givenName';
$CONF['attr_mapping']['mail']                   = 'mail';
$CONF['attr_mapping']['userPassword']           = 'userPassword';
// sort field to sort ldap users
$CONF['ldap_sort_field']                        = 'sn';
// sort order for ldap sort, allowed values are ASC and DESC
$CONF['ldap_sort_order']                        = 'ASC';
// use login data for ldap bind, allowed values are 0 and 1
$CONF['ldap_bind_use_login_data']                               = 0;
// LDAP bind dn suffix
$CONF['ldap_bind_dn_suffix']                                    = '';

//
// END OF CONFIG FILE
//
?>

searchrpc.php
<?php

/**
 * Ajax service for searches
 *
 * @author Thomas Krieger
 * @copyright 2008-2018 Thomas Krieger. All rights reserved.
 *           
 *            SVN Access Manager - a subversion access rights management tool
 *            Copyright (C) 2008-2018 Thomas Krieger <tom@svn-access-manager.org>
 *           
 *            This program is free software; you can redistribute it and/or modify
 *            it under the terms of the GNU General Public License as published by
 *            the Free Software Foundation; either version 2 of the License, or
 *            (at your option) any later version.
 *           
 *            This program is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU General Public License for more details.
 *           
 *            You should have received a copy of the GNU General Public License
 *            along with this program; if not, write to the Free Software
 *            Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *           
 *
 */

/*
 *
 * $LastChangedDate$
 * $LastChangedBy$
 *
 * $Id$
 *
 */
include ('load_config.php');

$installBase = isset($CONF[INSTALLBASE]) ? $CONF[INSTALLBASE] : "";

require ("$installBase/include/variables.inc.php");
include_once ("$installBase/include/constants.inc.php");
require_once ("$installBase/include/functions.inc.php");
require_once ("$installBase/include/db-functions-adodb.inc.php");
include_once ("$installBase/include/output.inc.php");

$error = 0;
$tErrorClass = "";
$tMessage = "";
$callback = isset($_GET['callback']) ? ($_GET['callback']) : "";
$maxRows = isset($_GET['maxRows']) ? ($_GET['maxRows']) : 10;
$filter = isset($_GET['name_startsWith']) ? ($_GET['name_startsWith']) : "";
$db = isset($_GET['db']) ? ($_GET['db']) : "";
$userid = isset($_GET['userid']) ? ($_GET['userid']) : "";
$user = isset($_GET['user']) ? ($_GET['user']) : "";
$group = isset($_GET['group']) ? ($_GET['group']) : "";
$project = isset($_GET['project']) ? ($_GET['project']) : "";
$admin = isset($_GET['admin']) ? ($_GET['admin']) : "";
$allowed = isset($_GET['allowed']) ? ($_GET['allowed']) : "";
$tArray = array();

list($ret, $SESSID_USERNAME ) = check_session_status();
if ($ret == 0) {
    
    $tArray[] = array(
            "name" => "Session expired!"
    );
    $data = json_encode($tArray);
}
elseif (strtolower($db) == "people") {
    
    $dbh = db_connect();
    $schema = db_determine_schema();
    $query = "SELECT id, name, givenname " . "  FROM " . $schema . "svnusers " . " WHERE ((userid like '%$filter%') " . "    OR  (CONCAT(givenname, ' ', name) like '%$filter%') " . "    OR  (CONCAT(name, ' ', givenname) like '%$filter%') " . "    OR  (name like '%$filter%') " . "    OR  (givenname like '%$filter%')) " . "   AND (deleted = '00000000000000') " . "ORDER BY name ASC, givenname ASC";
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        $data = array();
        if ($row[GIVENNAME] == "") {
            $data['name'] = $row['name'];
        }
        else {
            $data['name'] = $row[GIVENNAME] . " " . $row['name'];
        }
        $data['id'] = $row['id'];
        $tArray[] = $data;
    }
    
    db_disconnect($dbh);
}
elseif (strtolower($db) == "groups") {
    
    $dbh = db_connect();
    $rightAllowed = db_check_acl($SESSID_USERNAME, "Group admin", $dbh);
    $tGroupsAllowed = array();
    $schema = db_determine_schema();
    
    if ($rightAllowed == "none") {
        
        $tGroupsAllowed = db_check_group_acl($_SESSION[SVNSESSID]['username'], $dbh);
        if (count($tGroupsAllowed) == 0) {
            $groupAdmin = 2;
        }
        else {
            $groupAdmin = 1;
        }
    }
    else {
        $groupAdmin = 2;
    }
    
    if ($groupAdmin == 1) {
        
        $grouplist = "";
        
        foreach( $tGroupsAllowed as $groupid => $right ) {
            
            if ($grouplist == "") {
                $grouplist = "'" . $groupid . "'";
            }
            else {
                $grouplist .= ",'" . $groupid . "'";
            }
        }
        
        $grouplist = "(" . $grouplist . ")";
        
        $query = "SELECT  * " . "  FROM " . $schema . "svngroups " . " WHERE (deleted = '00000000000000') " . "   AND ((groupname like '%$filter%') " . "    OR (description like '%$filter%')) " . "   AND (id in $grouplist) " . "ORDER BY groupname ASC ";
    }
    else {
        
        $query = "SELECT id, groupname " . "  FROM " . $schema . "svngroups " . " WHERE ((groupname like '%$filter%') " . "    OR (description like '%$filter%')) " . "   AND (deleted = '00000000000000') " . "ORDER BY groupname ASC";
    }
    
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        $data = array();
        $data['name'] = $row['groupname'];
        $data['id'] = $row['id'];
        $tArray[] = $data;
    }
    
    db_disconnect($dbh);
}
elseif (strtolower($db) == "repos") {
    
    $dbh = db_connect();
    $schema = db_determine_schema();
    $query = "SELECT id, reponame " . "  FROM " . $schema . "svnrepos " . " WHERE ((repouser like '%$filter%') " . "    OR (reponame like '%$filter%')) " . "   AND (deleted = '00000000000000') " . "ORDER BY reponame ASC";
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        $data = array();
        $data['name'] = $row['reponame'];
        $data['id'] = $row['id'];
        $tArray[] = $data;
    }
    
    db_disconnect($dbh);
}
elseif (strtolower($db) == "projects") {
    
    $dbh = db_connect();
    $schema = db_determine_schema();
    $query = "SELECT * " . "  FROM " . $schema . "svnprojects " . " WHERE (svnmodule like '%$filter%') " . "   AND (deleted = '00000000000000') " . "ORDER BY svnmodule ASC";
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        $data = array();
        $data['name'] = $row['svnmodule'];
        $data['id'] = $row['id'];
        $tArray[] = $data;
    }
    
    db_disconnect($dbh);
}
elseif (strtolower($db) == "groupadmin") {
    
    $dbh = db_connect();
    $schema = db_determine_schema();
    $query = "SELECT svnusers.name, svnusers.givenname, svn_groups_responsible.id, svnusers.userid " . "  FROM " . $schema . "svn_groups_responsible," . $schema . "svnusers, " . $schema . "svngroups " . " WHERE (svn_groups_responsible.user_id = svnusers.id) " . "   AND (svnusers.deleted = '00000000000000') " . "   AND (svn_groups_responsible.deleted = '00000000000000') " . "   AND (svn_groups_responsible.group_id = svngroups.id) " . "   AND (svngroups.deleted = '00000000000000') " . "   AND ((svnusers.name like '%$filter%') " . "    OR  (svnusers.givenname like '%$filter%') " . "    OR  (svnusers.userid like '%$filter%') " . "    OR  (svngroups.groupname like '%$filter%') " . "    OR  (svngroups.description like '%$filter%')) " . "ORDER BY svnusers.name ASC, svnusers.givenname ASC";
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        $data = array();
        if ($row[GIVENNAME] != "") {
            $data['name'] = $row[GIVENNAME] . " " . $row['name'];
        }
        else {
            $data['name'] = $row['name'];
        }
        $data['id'] = $row['id'];
        $tArray[] = $data;
    }
    
    db_disconnect($dbh);
}
elseif (strtolower($db) == "accessright") {
    
    $dbh = db_connect();
    $schema = db_determine_schema();
    $tProjectIds = "";
    $query = "SELECT * " . "  FROM " . $schema . "svn_projects_responsible " . " WHERE (deleted = '00000000000000')";
    $result = db_query($query, $dbh);
    while ( $row = db_assoc($result[RESULT]) ) {
        
        if ($tProjectIds == "") {
            
            $tProjectIds = $row['project_id'];
        }
        else {
            
            $tProjectIds = $tProjectIds . "," . $row['project_id'];
        }
    }
    
    if ($tProjectIds != "") {
        
        $query = "SELECT svn_access_rights.id AS rid, svnmodule, modulepath, svnrepos." . "       reponame, valid_from, valid_until, path, access_right, `recursive`," . "       svn_access_rights.user_id, svn_access_rights.group_id, repopath " . "  FROM " . $schema . "svn_access_rights, " . $schema . "svnprojects, " . $schema . "svnrepos " . " WHERE (svnprojects.id = svn_access_rights.project_id) " . "   AND (svnprojects.id IN (" . $tProjectIds . "))" . "   AND (svnprojects.repo_id = svnrepos.id) " . "   AND (svn_access_rights.deleted = '00000000000000') " . "   AND ((svnmodule like '%$filter%') " . "    OR  (modulepath like '%$filter%') " . "    OR  (svnrepos.reponame like '%$filter%') " . "    OR  (path like '%$filter%') " . "    OR  (svnprojects.description like '%$filter%')) " . "ORDER BY svnrepos.reponame, svn_access_rights.path ";
        $result = db_query($query, $dbh);
        
        while ( $row = db_assoc($result[RESULT]) ) {
            
            $data = array();
            $data['name'] = $row['repopath'] . "|" . $row['path'] . "|" . $row['reponame'];
            $data['id'] = $row['rid'];
            $tArray[] = $data;
        }
    }
    
    db_disconnect($dbh);
}

$data = json_encode($tArray);
print $callback . "(" . $data . ");";

?>
